packages/rich-text/src/helpers/callbacks.ts

```diff
@@ -0,0 +1,32 @@
+import { toContentfulDocument } from '@contentful/contentful-slatejs-adapter';
+import { Document } from '@contentful/rich-text-types';
+import equal from 'fast-deep-equal';
+import debounce from 'lodash/debounce';
+
+import schema from '../constants/Schema';
+import { removeInternalMarks } from './removeInternalMarks';
+
+export const createOnChangeCallback = (handler?: (value: Document) => void) => {
+  // Cache previous value to avoid firing the handler unnecessarily
+  //
+  // Note: We are not using lodash/memoize here to avoid memory leaks
+  // due to having an infinite cache while we only care about the last
+  // value.
+  let cache: unknown = null;
+
+  return debounce((document: unknown) => {
+    if (equal(document, cache)) {
+      return;
+    }
+
+    cache = document;
+    const doc = removeInternalMarks(
```

Comment:
Where did this logic live before?

Reply:
It wasn't there. Looking at the existing implementation we used to have [useOnValueChanged](https://github.com/contentful/field-editors/blob/master/packages/rich-text/src/useOnValueChanged.ts) that did two things:
1. Skip selection changes. We don't want to trigger updates for those.
2. Avoid triggering `onChange` from incoming/external updates (via `onSkip` & `skip` which were passed in `RichTextEditor.tsx`)

To me, that's just a complicated way of saying: Don't trigger `onChange` if the actual JSON didn't change. Hence the change in this PR. If all things work out correctly then it should clean up the several `useState` and `useEffects` we had.